File: src/systems/Particles.ts
export interface Particle {
    x: number;
    y: number;
    vx: number;
    vy: number;
    life: number;
    maxLife: number;
    color: string;
    size: number;
}

export class ParticleSystem {
    particles: Particle[] = [];

    emit(x: number, y: number, count: number, speed: number, direction: number, spread: number, color: string, life: number) {
        for (let i = 0; i < count; i++) {
            const angle = direction + (Math.random() - 0.5) * spread;
            const velocity = speed * (0.5 + Math.random() * 0.5);

            this.particles.push({
                x,
                y,
                vx: Math.cos(angle) * velocity,
                vy: Math.sin(angle) * velocity,
                life: life * (0.8 + Math.random() * 0.4),
                maxLife: life, // Store original life if needed for fading
                color,
                size: Math.random() * 3 + 1
            });
        }
    }

    update(deltaTime: number) {
        for (let i = this.particles.length - 1; i >= 0; i--) {
            const p = this.particles[i];
            p.x += p.vx * deltaTime;
            p.y += p.vy * deltaTime;
            p.life -= deltaTime;

            if (p.life <= 0) {
                this.particles.splice(i, 1);
            }
        }
    }
}

File: python_project/particles.py
import math
import random

class Particle:
    def __init__(self, x, y, vx, vy, life, color, size):
        self.x = x
        self.y = y
        self.vx = vx
        self.vy = vy
        self.life = life
        self.maxLife = life
        self.color = color
        self.size = size

class ParticleSystem:
    def __init__(self):
        self.particles = []

    def emit(self, x, y, count, speed, angle, spread, color, life):
        for _ in range(count):
            a = angle + (random.random() - 0.5) * spread
            s = speed * (0.5 + random.random() * 0.5)
            vx = math.cos(a) * s
            vy = math.sin(a) * s
            size = random.random() * 2 + 1
            l = life * (0.5 + random.random() * 0.5)
            self.particles.append(Particle(x, y, vx, vy, l, color, size))

    def update(self, deltaTime):
        for p in self.particles:
            p.x += p.vx * deltaTime
            p.y += p.vy * deltaTime
            p.life -= deltaTime

        self.particles = [p for p in self.particles if p.life > 0]
